File: 0108-convert-sorted-array-to-binary-search-tree-将有序数组转换为二叉搜索树/0108.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        if nums:
            mid = int(len(nums) / 2)
            rtn = TreeNode(nums[mid])
            rtn.left, rtn.right = self.sortedArrayToBST(nums[:mid]), self.sortedArrayToBST(nums[mid+1:])
            return rtn

        

File: 0108-convert-sorted-array-to-binary-search-tree-将有序数组转换为二叉搜索树/0108.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
    TreeNode* build(vector<int>& nums, int beg, int end) {
        if (beg > end) return NULL;
        int mid = (beg + end) >> 1;
        TreeNode* root = new TreeNode(nums[mid]);
        root->left = build(nums, beg, mid - 1);
        root->right = build(nums, mid + 1, end);
        return root;
    }
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return build(nums, 0, nums.size() - 1);
    }
};


/*
class Solution {
private:
    TreeNode* helper(vector<int>& nums, int beg, int end) {
        if (end <= beg) return NULL;
        TreeNode* root = new TreeNode(0);
        int mid = (end + beg) / 2;
        root->val = nums[mid];
        root->left = helper(nums, beg, mid);
        root->right = helper(nums, mid+1, end);
        return root;
    }
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return helper(nums, 0, nums.size());
    }
};
*/